### PR TITLE
DIA-2895 GDPR consent expiration

### PR DESCRIPTION
**Pull request by andresilveirah**
* add `expirationDate` to response classes
* add `expirationDate` to `SPGDPRConsent` class
* on `.loadMessage`, set GDPR consent back to `.empty` if `exiprationDate < today`

CCPA will be implemented in a different PR.

### CODE REVIEW

File: ConsentViewController/Classes/Consents/SPGDPRConsent.swift [hunk 1]
```diff
@@ -156,20 +158,21 @@ public typealias SPGDPRPurposeId = String
...
+        euconsent = try container.decode(String.self, forKey: .euconsent)
+        tcfData = try container.decodeIfPresent(SPJson.self, forKey: .tcfData)
+        vendorGrants = try container.decode(SPGDPRVendorGrants.self, forKey: .vendorGrants)
+        childPmId = try container.decodeIfPresent(String.self, forKey: .childPmId)
+        consentStatus = try container.decode(ConsentStatus.self, forKey: .consentStatus)
+        webConsentPayload = try container.decodeIfPresent(SPWebConsentPayload.self, forKey: .webConsentPayload)
+        legIntCategories = try container.decodeIfPresent(Array.self, forKey: .legIntCategories)
+        legIntVendors = try container.decodeIfPresent(Array.self, forKey: .legIntVendors)
+        vendors = try container.decodeIfPresent(Array.self, forKey: .vendors)
+        categories = try container.decodeIfPresent(Array.self, forKey: .categories)
```
Comment (andresilveirah) [inline ConsentViewController/Classes/Consents/SPGDPRConsent.swift:172]: Simply removed `self` from the code above.

File: ConsentViewController/Classes/Extensions/Date.swift [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    var dayBefore: Date {
+        Calendar.current.date(byAdding: .day, value: -1, to: noon)!
+    }
+    var dayAfter: Date {
+        Calendar.current.date(byAdding: .day, value: 1, to: noon)!
+    }
+    var noon: Date {
+        Calendar.current.date(bySettingHour: 12, minute: 0, second: 0, of: self)!
+    }
+}
```
Comment (andresilveirah) [inline ConsentViewController/Classes/Extensions/Date.swift:24]: Helper methods, used only on tests, so far.

File: ConsentViewController/Classes/SourcePointClient/MessagesResponse.swift [hunk 3]
```diff
@@ -26,17 +26,10 @@ enum MessageCategory: Int, Codable, Defaultable, Equatable {
...
-
-        case .ios14:
-            return .ios14
-
-        default:
-            return .unknown
+        case .gdpr: .gdpr
+        case .ccpa: .ccpa
+        case .ios14: .ios14
+        default: .unknown
```
Comment (andresilveirah) [inline ConsentViewController/Classes/SourcePointClient/MessagesResponse.swift:32]: No changed logic, simply removing unnecessary `return` keyword and inlining things.

File: ConsentViewController/Classes/SourcePointClient/SourcepointClientCoordinator.swift [hunk 4]
```diff
@@ -699,25 +709,33 @@ class SourcepointClientCoordinator: SPClientCoordinator {
...
-                    self.state.gdpr?.euconsent = response.euconsent ?? getResponse?.gdpr?.euconsent ?? ""
-                    self.state.gdpr?.vendorGrants = response.grants ?? getResponse?.gdpr?.grants ?? SPGDPRVendorGrants()
-                    self.state.gdpr?.webConsentPayload = response.webConsentPayload ?? getResponse?.gdpr?.webConsentPayload
-                    self.state.gdpr?.legIntCategories = response.legIntCategories ?? getResponse?.gdpr?.legIntCategories
-                    self.state.gdpr?.legIntVendors = response.legIntVendors ?? getResponse?.gdpr?.legIntVendors
-                    self.state.gdpr?.vendors = response.vendors ?? getResponse?.gdpr?.vendors
-                    self.state.gdpr?.categories = response.categories ?? getResponse?.gdpr?.categories
-                    self.storage.spState = self.state
-
+                    self.handleGPDRPostChoice(action, getResponse, response)
```
Comment (andresilveirah) [inline ConsentViewController/Classes/SourcePointClient/SourcepointClientCoordinator.swift:738]: Moved the code above to new method `handleGDPRPostChoice`, for better readability.

File: ConsentViewController/Classes/SourcePointClient/MessagesResponse.swift (hunk 3 above)
Comment (bohdan-go-wombat) [reply ConsentViewController/Classes/SourcePointClient/MessagesResponse.swift:32]: I don't know swift, but can you omit the `return` keyword when you have to return a certain value?

Comment (andresilveirah) [reply ConsentViewController/Classes/SourcePointClient/MessagesResponse.swift:32]: @bohdan-go-wombat yes, if that's the only statement in a function. A switch statement _kinda_ returns a value after the `case :`, that's why it works in this case.

**Review by bohdan-go-wombat — APPROVED** [review]

File: Example/ConsentViewController_ExampleTests/SPClientCoordinator/SPClientCoordinatorSpec.swift [hunk 5]
```diff
@@ -461,6 +461,38 @@ class SPClientCoordinatorSpec: QuickSpec {
                     }
                 }
             }
+
+            describe("and expiration date is greater than current date") {
+                it("erases consent data and returns a message") {
+                    SPConsentManager.clearAllData()
```
Comment (carmelo-iriti) [inline Example/ConsentViewController_ExampleTests/SPClientCoordinator/SPClientCoordinatorSpec.swift:467]: Do you have only one method for deleting the consent? 
What’s the behaviour is for instance you have a list of 2 campaigns and one of them is expiring while the other not yet?

Comment (andresilveirah) [reply Example/ConsentViewController_ExampleTests/SPClientCoordinator/SPClientCoordinatorSpec.swift:467]: @carmelo-iriti the `.clearAllData` above is not related to consent expiration, but rather to clean all data before the test starts. 

The actual expiration logic is in the following lines: https://github.com/SourcePointUSA/ios-cmp-app/pull/505/files#diff-fc656d458a68144d63f7a40d9583ccf10a64bb53fd3cc75d0a30e13e8bca4a18R293

Keep in mind this is for GDPR only. For CCPA I implemented on #506

**Review by carmelo-iriti — APPROVED** [review]

File: ConsentViewController/Classes/SourcePointClient/MessagesResponse.swift (hunk 3 above)
Comment (andresilveirah) [reply ConsentViewController/Classes/SourcePointClient/MessagesResponse.swift]: @bohdan-go-wombat it turns out older versions of Swift complained about that 😅 , reverted back to have the `return` keyword.

**Pull request merged**